flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/BigQueryDynamicTableSource.java

```diff
@@ -0,0 +1,233 @@
+                        connectOptions.getTable())
+                .map(
+                        tuple -> {
+                            // we retrieve the existing partition ids and transform them into valid
+                            // values given the column data type
+                            return BigQueryPartition.partitionValuesFromIdAndDataType(
+                                            dataClient.retrieveTablePartitions(
+                                                    connectOptions.getProjectId(),
+                                                    connectOptions.getDataset(),
+                                                    connectOptions.getTable()),
+                                            tuple.f1)
+                                    .stream()
+                                    // for each of those valid partition values we create an map
+                                    // with the column name and the value
+                                    .map(
+                                            pValue -> {
+                                                Map<String, String> partitionColAndValue =
+                                                        new HashMap<>();
+                                                partitionColAndValue.put(tuple.f0, pValue);
+                                                return partitionColAndValue;
+                                            })
+                                    .collect(Collectors.toList());
+                        });
+    }
+
+    @Override
+    public void applyPartitions(List<Map<String, String>> remainingPartitions) {
+        this.readOptions =
+                this.readOptions
+                        .toBuilder()
+                        .setRowRestriction(
+                                // lets set the row restriction concating previously set restriction
+                                // (coming from the table definition) with the partition restriction
+                                // sent by Flink planner.
+                                this.readOptions.getRowRestriction()
+                                        + " AND "
+                                        + remainingPartitions.stream()
+                                                .flatMap(map -> map.entrySet().stream())
+                                                .map(
+                                                        entry ->
+                                                                "("
+                                                                        + entry.getKey()
+                                                                        + "="
+                                                                        + entry.getValue()
+                                                                        + ")")
+                                                .collect(Collectors.joining(" AND ")))
```

Comment:
[Needs fixing] Do we pushdown a filter on partition key here? Then, this should be an OR condition.

Reply:
Actually, it should AND different partition keys and OR the same keys.

Reply:
You are correct, partitions in BQ are disjoint sets. A record will be in only one partition in a table, so enforcing this with `AND` will make BQ to return empty sets.

Changed the `joining` statement and added another value on the tests to include 2 partitions so this surfaces in.